### PR TITLE
Nav: Add items to saved

### LINKED ISSUE #89531: Pin Nav Items: Implement way to pin each nav item in the Frontend
Figma exploration: [https://www.figma.com/proto/gATqMyVv1m94Sy0w6nXvMt/Explorations?page-id=0%3A1&type=[…]vMzcXajF8-1&scaling=min-zoom&starting-point-node-id=187%3A7693](https://www.figma.com/proto/gATqMyVv1m94Sy0w6nXvMt/Explorations?page-id=0%3A1&type=design&node-id=187-7693&viewport=-430%2C-2695%2C0.95&t=INW2KMTvMzcXajF8-1&scaling=min-zoom&starting-point-node-id=187%3A7693)

### PR DESCRIPTION
**Pull request by Clarity-89**
**What is this feature?**

Enable adding navigation items to saved. 

**Why do we need this feature?**

Part of the pinning nav items epic.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/89531

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/core/components/AppChrome/MegaMenu/MegaMenu.tsx [hunk 1]
```diff
@@ -46,6 +50,29 @@ export const MegaMenu = memo(
...
+        if (!id || !pinnedItems?.length) {
+          return false;
+        }
+        return pinnedItems?.includes(id);
+      },
+      [pinnedItems]
+    );
+
+    const onPinItem = (id?: string) => {
+      if (id) {
```
Comment (JoaoSilvaGrafana) [inline public/app/core/components/AppChrome/MegaMenu/MegaMenu.tsx:64]: Are there any menu items that don't have ids?

Comment (Clarity-89) [reply public/app/core/components/AppChrome/MegaMenu/MegaMenu.tsx:64]: haven't seen any but the id is marked as optional in the NavLinkItem type.

File: public/app/features/preferences/api.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    }),
+    updateUserPreferences: build.mutation<Preferences, UpdateRequestArg>({
+      query: ({ body }) => ({
+        url: URL,
+        method: 'PUT',
+        body,
+      }),
+      invalidatesTags: ['UserPreferences'],
+    }),
+    patchUserPreferences: build.mutation<Preferences, UpdateRequestArg>({
```
Comment (joshhunt) [inline public/app/features/preferences/api.ts:55]: Is this the correct type for PATCH? I would expect it allows you to update a partial of the user preferences

File: public/app/features/preferences/api.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  return backendSrvBaseQuery;
+}
+
+export interface UpdateRequestArg {
+  body: Preferences;
+}
+
+const URL = '/user/preferences';
```
Comment (joshhunt) [inline public/app/features/preferences/api.ts:37]: It looks like these endpoints are fully described in our OpenAPI spec, so I think we can generate this?

Edit `generate-rtk-apis.ts` to add a new output for `getUserPreferences` / `patchUserPreferences` / `updateUserPreferences` (and the org preferences as well?)

If we can generate it, then we don't have to have an opinion on the style of the output.

File: public/app/features/preferences/api/user/baseAPI.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+import { BackendSrvRequest, getBackendSrv } from '@grafana/runtime';
+
+interface RequestOptions extends BackendSrvRequest {
+  manageError?: (err: unknown) => { error: unknown };
+  showErrorAlert?: boolean;
+  body?: BackendSrvRequest['data'];
+}
+
+function createBackendSrvBaseQuery({ baseURL }: { baseURL: string }): BaseQueryFn<RequestOptions> {
```
Comment (Clarity-89) [inline public/app/features/preferences/api/user/baseAPI.ts:12]: I think we need to extract this somewhere a it's the same for a lot of the newly generated APIs.

Comment (joshhunt) [reply public/app/features/preferences/api/user/baseAPI.ts:12]: Yeah... Every RTKQ slice uses this same function, so really it should be common somewhere... Should be an easy change!

**Review by joshhunt — APPROVED** [review]

File: public/app/features/preferences/api/user/baseAPI.ts (hunk 4 above)
Comment (Clarity-89) [reply public/app/features/preferences/api/user/baseAPI.ts:12]: Yeah, but where should it be?

Comment (joshhunt) [reply public/app/features/preferences/api/user/baseAPI.ts:12]: idk, app/store? It doesn't have to be public, so I don't think it matters all that much.

I think it's something that can be tackeled in a seperate PR!

File: public/app/core/components/AppChrome/MegaMenu/hooks.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { useMemo } from 'react';
+
+import { config } from '@grafana/runtime';
+import { useGetUserPreferencesQuery } from 'app/features/preferences/api';
+
+export const usePinnedItems = () => {
+  const preferences = useGetUserPreferencesQuery();
+  const pinnedItems = useMemo(() => preferences.data?.navbar?.savedItemIds || [], [preferences]);
+
+  if (config.featureToggles.pinNavItems) {
```
Comment (JoaoSilvaGrafana) [inline public/app/core/components/AppChrome/MegaMenu/hooks.ts:10]: Nit: Should this `if` be at the start of the function so we don't make any unnecessary queries to the preferences endpoint?

**Review by JoaoSilvaGrafana — APPROVED** [review]

File: public/app/core/components/AppChrome/MegaMenu/hooks.ts (hunk 5 above)
Comment (Clarity-89) [reply public/app/core/components/AppChrome/MegaMenu/hooks.ts:10]: Hooks can't be called conditionall, we'd need to add a condition inside the API call itself. I don't think it's a big deal as it only fetches org preferences and the feature toggle is temporary anyways.

**Pull request merged**